src/background/providers/cloudflare.test.ts

```diff
@@ -116,17 +215,58 @@ describe('issuance', () => {
             );
         });
 
-        /*
-         * The request is invalid if any of the followings is true:
-         * 1. It has no url param of any of the followings:
-         *    a. '__cf_chl_captcha_tk__'
-         *    b. '__cf_chl_managed_tk__'
-         * 2. It has no body param of any of the followings:
-         *    a. 'g-recaptcha-response'
-         *    b. 'h-captcha-response'
-         *    c. 'cf_captcha_kind'
-         */
-        test('invalid request', async () => {
+        test('with issueInfo without Referer header', async () => {
+            const storage = new StorageMock();
+            const updateIcon = jest.fn();
+            const navigateUrl = jest.fn();
+
+            const provider = new CloudflareProvider(storage, { updateIcon, navigateUrl });
+            const tokens = [new Token(), new Token(), new Token()];
+            const issue = jest.fn(async () => {
+                return tokens;
+            });
+            provider['issue'] = issue;
+            const issueInfo = {
+                requestId: 'xxx',
+                formData: {
+                    ['h-captcha-response']: 'body-param',
+                    ['cf_captcha_kind']: 'body-param',
+                },
+            };
+            provider['issueInfo'] = issueInfo;
+            const details = {
+                method: 'POST',
+                url: 'https://captcha.website/?__cf_chl_f_tk=token',
+                requestId: 'xxx',
+                frameId: 1,
+                parentFrameId: 1,
+                tabId: 1,
+                type: 'xmlhttprequest' as chrome.webRequest.ResourceType,
+                timeStamp: 1,
+                requestHeaders: [],
+            };
+            const result = await provider.handleBeforeSendHeaders(details);
+            expect(result).toStrictEqual({ cancel: true });
+            const newIssueInfo = provider['issueInfo'];
+            expect(newIssueInfo).toBeNull();
+
+            expect(issue.mock.calls.length).toBe(1);
+            expect(issue).toHaveBeenCalledWith('https://captcha.website/?__cf_chl_f_tk=token', {
```

Comment:
Should we check that this request's `Referer` header will either not be present or contain the URL without the parameter?

Reply:
I meant for the challenge solve request, and the `Referer` not being present is perfect. Thanks!